File: native/apple/Sources/IosAwnCore/AwesomeEventsReceiver.swift
import Foundation

/// Broadcaster the engine uses to deliver lifecycle events to its subscribers
/// (the Flutter plugin). Mirrors the original IosAwnCore `AwesomeEventsReceiver`.
public final class AwesomeEventsReceiver {
    public static let shared = AwesomeEventsReceiver()
    private init() {}

    private var listeners: [AwesomeEventListener] = []

    @discardableResult
    public func subscribe(listener: AwesomeEventListener) -> AwesomeEventsReceiver {
        if !listeners.contains(where: { $0 === listener }) {
            listeners.append(listener)
        }
        return self
    }

    public func unsubscribe(listener: AwesomeEventListener) {
        listeners.removeAll { $0 === listener }
    }

    func notifyAwesomeEvent(eventType: String, content: [String: Any]) {
        for listener in listeners {
            listener.onNewAwesomeEvent(eventType: eventType, content: content)
        }
    }
}

File: native/android/AndroidAwnCore/src/main/kotlin/me/carda/android_awn_core/AwesomeEventsReceiver.kt
package me.carda.android_awn_core

/**
 * Static broadcaster the engine (and the dismiss BroadcastReceiver, which lives
 * outside the plugin instance) use to deliver lifecycle events to subscribers
 * (the Flutter plugin). Mirrors the original AndroidAwnCore `AwesomeEventsReceiver`.
 */
object AwesomeEventsReceiver {
    private val listeners = mutableListOf<AwesomeEventListener>()

    fun subscribe(listener: AwesomeEventListener) {
        if (!listeners.contains(listener)) listeners.add(listener)
    }

    fun unsubscribe(listener: AwesomeEventListener) {
        listeners.remove(listener)
    }

    fun notifyAwesomeEvent(eventType: String, content: Map<String, Any?>) {
        for (listener in ArrayList(listeners)) {
            listener.onNewAwesomeEvent(eventType, content)
        }
    }
}
